forte/data/ontology/core.py

```diff
@@ -446,30 +408,33 @@ class FNdArray:
     """
 
     def __init__(
-        self, dtype: Optional[str] = None, shape: Optional[List[int]] = None
+        self,
+        dtype: Optional[str] = None,
+        shape: Optional[List[int]] = None,
+        data_ref: Optional[List] = None,
     ):
         super().__init__()
-        self._dtype: Optional[np.dtype] = (
-            np.dtype(dtype) if dtype is not None else dtype
+        self.__data_ref: List = (
```

Comment:
looks like we convert the 3 values into one list `data_ref`? Any reasons for this?

Also, it seems `dtype` and `shape` contain the same information as `data_ref` in the arguments.

Reply:
Yes. The reason is to store only single source of truth for `FNdArray` inside `DataStore`. Each `FNdArray` attribute will be stored as a list of `data_ref=[dtype, shape, data]`. All the getter and setter method of `FNdArray` object should now be directed into the reference of that list in `DataStore`.

`dtype` and `shape` are still kept for backward compatibility.

Reply:
a single source of truth internally is fine. But do we need to change the interface from separate attributes `dtype` and `shape` to `data_ref`? 

During usage, a user would still prefer the more meaningful full `shape` rather than an ambiguous `data_ref`. 

By using kwargs, I actually mean when we are calling the constructor, we should can pass all arguments as kwargs to the constructor, not adding one additional kwargs, i.e. doing something like `FNdArray(**data_ref)`, which is simply `FNdArray(**{shape=x, dtype=y})`

Reply:
basically, we don't need to introduce one additional `kwargs` or `data_ref`. Notice `FNdArray` is already a user-facing interface, which is something we should keep clean and logical

Reply:
Then maybe it's better to keep the interface as `FNdArray(shape=x, dtype=y)`. Meanwhile we can also add a new internal method called `FNdArray.set_data_ref(data_ref)` to pass the reference.

Reply:
what is the reference here? I only saw `self.__data_ref = [dtype, shape, None]` in the constructor here.

Reply:
It's also a list of `[dtype, shape, data]` but it's stored inside `DataStore` and can be retrieved with `DataStore.get_attributes`. `set_data_ref` will overwrite the list created in constructor
```python
def set_data_ref(data_ref):
    self. __data_ref = data_ref
```
so that `self. __data_ref` is now linked to a list stored in `DataStore`.